File: session-notes/cleanup-notes/cleanup_session_notes.py
# Python code for the specified tasks

import re
import os
import shutil
import sys

def get_date_from_filename(filename):
    match = re.search(r'(\d{4}-\d{2}-\d{2})', filename, re.IGNORECASE)
    return match.group(1) if match else None

def slurp_file(filename):
    try:
        with open(filename, 'r') as file:
            return file.read()
    except FileNotFoundError:
        return None

def contains_inactive_coauthors(contents):
    return bool(re.search(r'^#+\s*Inactive Co-Authors', contents, re.IGNORECASE | re.MULTILINE))

def contains_active_coauthors(contents):
    return bool(re.search(r'^#+\s*Active Co-Authors', contents, re.IGNORECASE | re.MULTILINE))

def contains_session_date(contents):
    return bool(re.search(r'^#+\s*Session Date', contents, re.IGNORECASE | re.MULTILINE))

def delete_inactive_coauthors(contents):
    return re.sub(r'^#+\s*Inactive Co-Authors.*?(?=^#|\Z)', '', contents, flags=re.IGNORECASE | re.MULTILINE | re.DOTALL)

def normalize_coauthor_heading(contents):
    return re.sub(r'^#+\s*.*Co-Author.*', '## Co-Authors', contents, flags=re.IGNORECASE | re.MULTILINE)

def cleanup_file(filename):
    original_contents = slurp_file(filename)
    if original_contents is None:
        print(f"File not found: {filename}")
        return

    contents = original_contents
    date_as_string = get_date_from_filename(filename)

    if not contains_session_date(contents):
        contents = f"# Session Date: {date_as_string}\n" + contents

    if contains_active_coauthors(contents) and contains_inactive_coauthors(contents):
        contents = delete_inactive_coauthors(contents)

    contents = normalize_coauthor_heading(contents)

    if contents == original_contents:
        print(f"No changes were needed for the file: {filename}")
    else:
        print(f"Changes were made to the file: {filename}")
        new_filename = filename + ".new"
        with open(new_filename, 'w') as new_file:
            new_file.write(contents)
        shutil.copystat(filename, new_filename)
        original_backup_filename = filename + ".original"
        os.rename(filename, original_backup_filename)
        os.rename(new_filename, filename)
        print(f"You can view changes by issuing this command: diff -u {original_backup_filename} {filename}")

def main():
    for filename in sys.argv[1:]:
        cleanup_file(filename)

if __name__ == "__main__":
    main()

File: session-notes/cleanup-notes/cleanup_session_notes.pl
use strict;
use warnings;

# Sub-task: Create subroutine 'get_date_from_filename'
sub get_date_from_filename {
    my ($filename) = @_;
    $filename =~ /(\d{4}-\d{2}-\d{2})/;
    return $1;
}

# Sub-task: Create subroutine 'slurp_file'
sub slurp_file {
    my ($filename) = @_;
    local $/;
    open my $fh, '<', $filename or die "Could not open file '$filename': $!";
    my $contents = <$fh>;
    close $fh;
    return $contents;
}

# Sub-task: Create subroutine 'contains_inactive_coauthors'
sub contains_inactive_coauthors {
    my ($contents) = @_;
    return $contents =~ /#+\s*Inactive Co-Authors/i;
}

# Sub-task: Create subroutine 'contains_active_coauthors'
sub contains_active_coauthors {
    my ($contents) = @_;
    return $contents =~ /#+\s*Active Co-Authors/i;
}

# Sub-task: Create subroutine 'contains_session_date'
sub contains_session_date {
    my ($contents) = @_;
    return $contents =~ /#+\s*Session Date/i;
}

# Sub-task: Create subroutine 'delete_inactive_coauthors'
sub delete_inactive_coauthors {
    $_[0] =~ s/(#+\s*Inactive Co-Authors.*?)(?=#+\s*\w)/\n/si;
}

# Sub-task: Create subroutine 'normalize_coauthor_heading'
sub normalize_coauthor_heading {
    $_[0] =~ s/#+\s*.*?Co-Author.*/## Co-Authors/gi;
}

# Sub-task: Create subroutine 'cleanup_file'
sub cleanup_file {
    my ($filename) = @_;
    my $original_contents = slurp_file($filename);
    my $contents = $original_contents;
    my $date_as_string = get_date_from_filename($filename);

    # Add session date if not present
    $contents = "## Session Date: $date_as_string\n\n$contents" unless contains_session_date($contents);

    # Delete inactive coauthors if necessary
    delete_inactive_coauthors($contents) if contains_active_coauthors($contents) && contains_inactive_coauthors($contents);

    # Normalize coauthor headings
    normalize_coauthor_heading($contents);

    if ($contents eq $original_contents) {
        print "No changes were needed for $filename\n";
    } else {
        print "Changes were made to $filename\n";
        my ($mode, $atime, $mtime) = (stat $filename)[2, 8, 9];
        my $new_filename = $filename . '.new';
        open my $fh, '>', $new_filename or die "Could not open file '$new_filename': $!";
        print $fh $contents;
        close $fh;
        chmod $mode & 07777, $new_filename;
        utime $atime, $mtime, $new_filename;
        
        # File manipulation
        rename $filename, $filename . '.original';
        rename $new_filename, $filename;

        print "You can view changes by issuing this command:\n";
        print "diff -u " . $filename . ".original " . $filename . "\n";
    }
}

# Sub-task: Create subroutine 'main'
sub main {
    foreach my $filename (@ARGV) {
        cleanup_file($filename);
    }
}

# Call main
main() if @ARGV;
